File: Tools/Simulation/StoffGo/shapes2D/boundary.go
package shapes2d

import (
	"math"
	G "stoffgo/geometry"
	"stoffgo/tools"
)

type Boundary struct {
	start       int
	end         int
	orientation bool
}

func (b *PhysicsBox) pointInBoundary(pt G.Vec2) bool {
	tools.Assert(len(b.Boundaries) > 0, "We need a boundary before calling Hex")
	BoundarySimPoints := b.Points[b.Boundaries[0].start:b.Boundaries[0].end]
	Boundary := make([]G.Vec2, len(BoundarySimPoints))
	for i, v := range BoundarySimPoints {
		Boundary[i] = v.pos
	}

	return G.IsPointInPolygon(pt, Boundary)
}

func (p *Plane) pointInBoundary(pt G.Vec2) bool {
	tools.Assert(len(p.Boundaries) > 0, "We need a boundary before calling Hex")
	return G.IsPointInPolygon(pt, p.Vertices[p.Boundaries[0].start:p.Boundaries[0].end]) ||
		G.IsPointOnPolygon(pt, p.Vertices[p.Boundaries[0].start:p.Boundaries[0].end], 0.0000000001)
}

func normalizeCurve(v []G.Vec2, target_pt_distance float64) []G.Vec2 {
	var newVertices []G.Vec2

	v = append(v, v[0])
	newVertices = append(newVertices, v[0])
	n := len(v)

	currentDistance := 0.0

	for i := 0; i < n; i++ {
		nextDistance := v[(i+1)%n].Distance(v[i])
		if currentDistance+nextDistance < target_pt_distance {
			continue
		}

		if currentDistance > 0 {
			currentDistance = 0
			newVertices = append(newVertices, v[i])
			continue
		}

		dirVec := v[(i+1)%n].Sub(v[i])
		dir := dirVec.Normalize()
		amt := int(math.Ceil(nextDistance / target_pt_distance))

		partition_len := dirVec.Length() / float64(amt)

		for j := 0; j < amt; j++ {
			newPoint := v[i].Add(dir.Scale(float64(j+1) * partition_len))
			newVertices = append(newVertices, newPoint)
		}

		currentDistance = 0
	}

	// Implementing the final if block
	first := newVertices[0]
	last := newVertices[len(newVertices)-1]
	secondLast := newVertices[len(newVertices)-2]

	if last.Distance(first) < 5 {
		if secondLast.Distance(first) < 10 {
			// Remove the last point
			newVertices = newVertices[:len(newVertices)-1]
		} else {
			// Move the last point to halfway between the first and second last point
			newVertices[len(newVertices)-1] = secondLast.Add(first).Scale(0.5)
		}
	}

	return newVertices
}

func computeBB(vertices []G.Vec2) [2]G.Vec2 {
	min := G.Vec2{math.Inf(1), math.Inf(1)}
	max := G.Vec2{math.Inf(-1), math.Inf(-1)}

	for _, v := range vertices {
		if v[0] < min[0] {
			min[0] = v[0]
		}
		if v[1] < min[1] {
			min[1] = v[1]
		}
		if v[0] > max[0] {
			max[0] = v[0]
		}
		if v[1] > max[1] {
			max[1] = v[1]
		}
	}

	return [2]G.Vec2{min, max}
}

File: Sewing/Tools/Simulation/StoffPy/boundary.py
import numpy as np
import math

class Boundary:
    def __init__(self, box, start, end, orientation):
        self.start = start
        self.end = end
        self.orientation = orientation
        self.box = box

def normalize_curve(v, target_pt_distance):
    # Initialize the new vertices list with the first vertex
    new_vertices = [v[0]]

    # Append the first vertex to the end of the array to form a closed loop
    v = np.append(v, [v[0]], axis=0)
    n = len(v)

    current_distance = 0.0

    for i in range(n - 1):
        next_distance = np.linalg.norm(v[(i + 1) % n] - v[i])
        if current_distance + next_distance < target_pt_distance:
            continue

        if current_distance > 0:
            current_distance = 0
            new_vertices.append(v[i])
            continue

        dir_vec = v[(i + 1) % n] - v[i]
        dir = dir_vec / np.linalg.norm(dir_vec)
        amt = int(math.ceil(next_distance / target_pt_distance))

        partition_len = np.linalg.norm(dir_vec) / amt

        for j in range(amt):
            new_point = v[i] + dir * (j + 1) * partition_len
            new_vertices.append(new_point)

        current_distance = 0

    # Convert the list of new vertices back to a NumPy array
    new_vertices = np.array(new_vertices)

    # Implementing the final if block
    first = new_vertices[0]
    last = new_vertices[-1]
    second_last = new_vertices[-2]

    if np.linalg.norm(last - first) < 5:
        if np.linalg.norm(second_last - first) < 10:
            # Remove the last point
            new_vertices = new_vertices[:-1]
        else:
            # Move the last point to halfway between the first and second last point
            new_vertices[-1] = (second_last + first) * 0.5

    return new_vertices
